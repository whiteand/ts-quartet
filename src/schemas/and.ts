import { IAndSchema, TSchema } from "../types";
import { SchemaType } from "./SchemaType";

export function and(schemas: TSchema[]): IAndSchema {
  return {
    schemas,
    type: SchemaType.And
  };
}
