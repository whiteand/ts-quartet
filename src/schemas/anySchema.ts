import { IAnySchema } from "../types";
import { SchemaType } from "./SchemaType";

const ANY_SCHEMA: IAnySchema = {
  type: SchemaType.Any,
};

export function anySchema(): IAnySchema {
  return ANY_SCHEMA;
}
