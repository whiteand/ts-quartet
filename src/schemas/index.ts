export * from "./and";
export * from "./anySchema";
export * from "./array";
export * from "./arrayOf";
export * from "./boolean";
export * from "./custom";
export * from "./finite";
export * from "./functionSchema";
export * from "./max";
export * from "./maxLength";
export * from "./min";
export * from "./minLength";
export * from "./negative";
export * from "./neverSchema";
export * from "./not";
export * from "./notANumber";
export * from "./number";
export * from "./objectSchema";
export * from "./pair";
export * from "./positive";
export * from "./safeInteger";
export * from "./SchemaType";
export * from "./SpecialProp";
export * from "./string";
export * from "./symbol";
export * from "./testSchema";
export * from "./variant";
