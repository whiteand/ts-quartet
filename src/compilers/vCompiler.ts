import { RawSchema, rawSchemaToSchema } from "../rawSchemaToSchema";
import { vSchemaRenderer } from "../renderers/v";
import { SchemaType } from "../schemas";
import {
  Alloc,
  CompilationResult,
  ICompilationResultProps,
  TSchema,
  Validator
} from "../types";
import { arrToDict, getKeyAccessor, has } from "../utils";

const VALUE_PARAM_NAME = "value";
const CTX_PARAM_NAME = "ctx";
type Ctx = ICompilationResultProps<any> & {
  memory: Record<string, any>;
  pure: EvaluatedFunc;
};
type EvaluatedFunc = (value: any, ctx: Ctx) => boolean;

const SINGLE_EXPRESSION_SCHEMA_TYPE: Partial<
  Record<SchemaType, boolean>
> = arrToDict([
  SchemaType.Any,
  SchemaType.Array,
  SchemaType.Boolean,
  SchemaType.Finite,
  SchemaType.Function,
  SchemaType.Max,
  SchemaType.MaxLength,
  SchemaType.Min,
  SchemaType.MinLength,
  SchemaType.Negative,
  SchemaType.Never,
  SchemaType.Number,
  SchemaType.Positive,
  SchemaType.SafeInteger,
  SchemaType.String,
  SchemaType.Symbol
]);

function getFunctionBodyCode(schema: TSchema, ctx: Ctx) {
  const alloc: Alloc = (initialValue = null, prefix = "t") => {
    if (!has(ctx.memory, prefix)) {
      ctx.memory[prefix] = initialValue;
      return `${CTX_PARAM_NAME}.memory${getKeyAccessor(prefix)}`;
    }
    let i = 2;
    while (true) {
      const key = prefix + i;
      if (!has(ctx.memory, key)) {
        ctx.memory[key] = initialValue;
        return `${CTX_PARAM_NAME}.memory${getKeyAccessor(key)}`;
      }
      i++;
    }
  };
  if (
    !schema ||
    typeof schema !== "object" ||
    SINGLE_EXPRESSION_SCHEMA_TYPE[schema.type] === true
  ) {
    return `
        return ${vSchemaRenderer.getExpr(
          VALUE_PARAM_NAME,
          schema,
          alloc,
          `${CTX_PARAM_NAME}.memory.pathToValueId`,
          `${CTX_PARAM_NAME}.explanations`,
          vCompileSchema
        )}
      `.trim();
  }
  if (
    schema &&
    typeof schema === "object" &&
    schema.type === SchemaType.Variant
  ) {
    return `
${vSchemaRenderer.getIfExprReturnTrue(
  VALUE_PARAM_NAME,
  schema,
  alloc,
  `${CTX_PARAM_NAME}.memory.pathToValueId`,
  `${CTX_PARAM_NAME}.explanations`,
  vCompileSchema
)}
return false
    `.trim();
  }
  return `
${vSchemaRenderer.getIfNotExprReturnFalse(
  VALUE_PARAM_NAME,
  schema,
  alloc,
  `${CTX_PARAM_NAME}.memory.pathToValueId`,
  `${CTX_PARAM_NAME}.explanations`,
  vCompileSchema
)}
return true
`.trim();
}

function vCompileSchema<T = any>(
  schema: TSchema
): CompilationResult<T, any> & Ctx {
  const ctx: Ctx = {
    explanations: [],
    memory: {
      pathToValueId: []
    },
    pure: () => false
  };

  ctx.pure = new Function(
    VALUE_PARAM_NAME,
    CTX_PARAM_NAME,
    getFunctionBodyCode(schema, ctx)
  ) as EvaluatedFunc;

  const validator = function(value: any) {
    return ctx.pure(value, ctx);
  } as Validator<T>;

  return Object.assign(validator, ctx);
}

export function vCompiler<T = any>(
  rawSchema: RawSchema
): CompilationResult<T, any> & Ctx {
  const schema = rawSchemaToSchema(rawSchema);
  return vCompileSchema<T>(schema);
}
