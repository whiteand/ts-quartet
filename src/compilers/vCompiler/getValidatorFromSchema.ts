import { Validator, TSchema } from "../../types";
import { validate } from "./validate";
import { SchemaType } from "../../schemas/SchemaType";
import { ifInvalidReturnFalse } from "./ifInvalidReturnFalse";
import { getAlloc } from "./getAlloc";
import { arrayOf } from "../../schemas";

export function getValidatorFromSchema(
  schema: TSchema,
  key: string | number | undefined
): (value: any) => boolean {
  if (typeof schema !== "object" || schema === null) {
    return (value: any) => value === schema;
  }
  switch (schema.type) {
    case SchemaType.And:
    case SchemaType.Object:
    case SchemaType.ArrayOf:
      const context: Record<string, any> = {};
      const contextParamName = "ctx";
      const alloc = getAlloc(context, contextParamName);
      const funcBody = `
        ${ifInvalidReturnFalse(schema, alloc, "value", key)}
        return true
      `;
      const isValid = new Function("value", contextParamName, funcBody) as (
        value: any,
        context: any
      ) => boolean;
      return (value) => {
        try {
          return isValid(value, context);
        } catch (error) {
          console.debug(isValid.toString());
          throw error;
        }
      };
    case SchemaType.Any:
      return () => true;
    case SchemaType.Array:
      return (value) => Array.isArray(value);
    case SchemaType.Boolean:
      return (value) => typeof value === "boolean";
    case SchemaType.Finite:
      return (value) => Number.isFinite(value);
    case SchemaType.Function:
      return (value) => typeof value === "function";
    case SchemaType.Max:
      return schema.isExclusive
        ? (value) => value < schema.maxValue
        : (value) => value <= schema.maxValue;
    case SchemaType.MaxLength:
      return schema.isExclusive
        ? (value) => value != null && value.length < schema.maxLength
        : (value) => value != null && value.length <= schema.maxLength;
    case SchemaType.Min:
      return schema.isExclusive
        ? (value) => value > schema.minValue
        : (value) => value >= schema.minValue;
    case SchemaType.MinLength:
      return schema.isExclusive
        ? (value) => value != null && value.length > schema.minLength
        : (value) => value != null && value.length >= schema.minLength;
    case SchemaType.Negative:
      return (value) => value < 0;
    case SchemaType.Never:
      return () => false;
    case SchemaType.Not:
      const isNotValid = getValidatorFromSchema(schema.schema, key);
      return (value) => !isNotValid(value);
    case SchemaType.NotANumber:
      return (value) => Number.isNaN(value);
    case SchemaType.Number:
      return (value) => typeof value === "number";
    case SchemaType.Pair:
      const isValidPair = getValidatorFromSchema(schema.keyValueSchema, key);
      let pair = {
        value: undefined,
        key,
      };
      return (value) => {
        pair.value = value;
        return isValidPair(pair);
      };
    case SchemaType.Positive:
      return (value) => value > 0;
    case SchemaType.SafeInteger:
      return (value) => Number.isSafeInteger(value);
    case SchemaType.String:
      return (value) => typeof value === "string";
    case SchemaType.Symbol:
      return (value) => typeof value === "symbol";
    case SchemaType.Test:
      return (value) => schema.tester.test(value);
    case SchemaType.Custom:
      return schema.customValidator;
    case SchemaType.Variant:
      const { variants } = schema;
      const compiledVariants = variants.map((innerSchema) =>
        getValidatorFromSchema(innerSchema, key)
      );
      return (value) => {
        for (let i = 0; i < compiledVariants.length; i++) {
          const compiledVariant = compiledVariants[i];
          if (compiledVariant(value)) return true;
        }
        return false;
      };
  }
}
