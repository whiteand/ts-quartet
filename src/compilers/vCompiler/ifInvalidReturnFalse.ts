import { SchemaType } from "../../schemas";
import { TSchema, Z } from "../../types";
import { getAccessorWithAlloc, has } from "../../utils";
import { getValidatorFromSchema } from "./getValidatorFromSchema";

function fromNegation(negationConditionCode: string) {
  return `if (${negationConditionCode}) return false`;
}

export function ifInvalidReturnFalse(
  schema: TSchema,
  alloc: (varName: string, initialValue: Z, singleton?: boolean) => string,
  valueAddress: string,
  keyAddress: string | number | undefined,
): string {
  if (schema === null) {
    return fromNegation(`${valueAddress} !== null`);
  }
  if (schema === undefined) {
    return fromNegation(`${valueAddress} !== undefined`);
  }
  if (typeof schema !== "object") {
    const primitiveValueAddress = alloc("primitive", schema);
    return fromNegation(`${valueAddress} !== ${primitiveValueAddress}`);
  }
  switch (schema.type) {
    case SchemaType.And: {
      const ifInvalidReturnFalseStatements = schema.schemas.map((innerSchema) =>
        ifInvalidReturnFalse(innerSchema, alloc, valueAddress, keyAddress),
      );
      return ifInvalidReturnFalseStatements.join("\n");
    }
    case SchemaType.Any:
      return "";
    case SchemaType.Array:
      return fromNegation(`!Array.isArray(${valueAddress})`);
    case SchemaType.ArrayOf: {
      const { elementSchema } = schema;
      const incrementVar = alloc("i", 0);
      const elementVar = alloc("e", undefined);
      const checkIsArray = `if (!Array.isArray(${valueAddress})) return false;`;
      const checkElements = `
        for (${incrementVar} = 0; ${incrementVar} < ${valueAddress}.length; ${incrementVar}++) {
            ${elementVar} = ${valueAddress}[${incrementVar}];
            ${ifInvalidReturnFalse(
              elementSchema,
              alloc,
              elementVar,
              incrementVar,
            )}
        }
      `;
      return [checkIsArray, checkElements].join("\n");
    }
    case SchemaType.Boolean:
      return fromNegation(`typeof ${valueAddress} !== 'boolean'`);
    case SchemaType.Finite:
      return fromNegation(`!Number.isFinite(${valueAddress})`);
    case SchemaType.Function:
      return fromNegation(`typeof ${valueAddress} !== 'function'`);
    case SchemaType.Max: {
      const maxValueVar = alloc("maxValue", schema.maxValue);
      const cmpMax = schema.isExclusive ? "<" : "<=";
      return fromNegation(`!(${valueAddress} ${cmpMax} ${maxValueVar})`);
    }
    case SchemaType.MaxLength: {
      const maxLengthVar = alloc("maxLength", schema.maxLength);
      const cmpMaxLength = schema.isExclusive ? "<" : "<=";
      return fromNegation(
        `${valueAddress} == null || !(${valueAddress}.length ${cmpMaxLength} ${maxLengthVar})`,
      );
    }
    case SchemaType.Min: {
      const minValueVar = alloc("minValue", schema.minValue);
      const cmpMin = schema.isExclusive ? ">" : ">=";
      return fromNegation(`!(${valueAddress} ${cmpMin} ${minValueVar})`);
    }
    case SchemaType.MinLength: {
      const minLengthVar = alloc("minLength", schema.minLength);
      const cmp = schema.isExclusive ? ">" : ">=";
      return fromNegation(
        `${valueAddress} == null || !(${valueAddress}.length ${cmp} ${minLengthVar})`,
      );
    }
    case SchemaType.Negative:
      return fromNegation(`!(${valueAddress} < 0)`);
    case SchemaType.Never:
      return "return false";
    case SchemaType.Not: {
      if (typeof schema.schema !== "object" || schema.schema === null) {
        if (schema.schema === null) {
          return fromNegation(`${valueAddress} === null`);
        }
        if (schema.schema === undefined) {
          return fromNegation(`${valueAddress} === undefined`);
        }
        const primitiveValueAddress = alloc("primitive", schema.schema);
        return fromNegation(`${valueAddress} === ${primitiveValueAddress}`);
      }
      const isValid = getValidatorFromSchema(schema.schema, keyAddress);
      const isValidVar = alloc("isValid", isValid);
      return `if (${isValidVar}(${valueAddress})) return false`;
    }
    case SchemaType.NotANumber:
      return fromNegation(`!Number.isNaN(${valueAddress})`);
    case SchemaType.Number:
      return fromNegation(`typeof ${valueAddress} !== 'number'`);
    case SchemaType.Object: {
      const statements: string[] = [
        `if (${valueAddress} == null) return false`,
      ];
      for (let i = 0; i < schema.props.length; i++) {
        const prop = schema.props[i];
        const propSchema = schema.propsSchemas[prop];
        const propAccessor = getAccessorWithAlloc(prop, alloc);
        const checkPropStatement = ifInvalidReturnFalse(
          propSchema,
          alloc,
          `${valueAddress}${propAccessor}`,
          JSON.stringify(prop),
        );
        statements.push(checkPropStatement);
      }
      if (schema.hasRestValidator) {
        const restOmitDictVar = alloc("ro", schema.restOmitDict);
        const restPropsVar = alloc("r", []);
        const incVar = alloc("i", 0);
        statements.push(`${restPropsVar} = Object.keys(${valueAddress})`);
        statements.push(
          `for (${incVar} = 0; ${incVar} < ${restPropsVar}.length; ${incVar}++) {`,
        );
        const restPropVar = alloc("rp", undefined);
        statements.push(`${restPropVar} = ${restPropsVar}[${incVar}];`);
        const propsSchemasVar = alloc("ps", schema.propsSchemas);
        const hasVar = alloc("has", has, true);
        statements.push(
          `if (${hasVar}(${propsSchemasVar}, ${restPropVar}) || ${restOmitDictVar}[${restPropVar}] === true) continue;`,
        );
        const restValueVar = alloc("rv", undefined);
        statements.push(`${restValueVar} = ${valueAddress}[${restPropVar}]`);
        statements.push(
          ifInvalidReturnFalse(schema.rest, alloc, restValueVar, restPropVar),
        );
        statements.push("}");
      }
      return statements.join("\n");
    }
    case SchemaType.Pair: {
      const pairVar = alloc("pair", { value: undefined, key: undefined });
      return [
        `${pairVar}.value = ${valueAddress}`,
        `${pairVar}.key = ${keyAddress}`,
        `${ifInvalidReturnFalse(
          schema.keyValueSchema,
          alloc,
          pairVar,
          keyAddress,
        )}`,
      ].join("\n");
    }
    case SchemaType.Positive:
      return fromNegation(`!(${valueAddress} > 0)`);
    case SchemaType.SafeInteger:
      return fromNegation(`!Number.isSafeInteger(${valueAddress})`);
    case SchemaType.String:
      return fromNegation(`typeof ${valueAddress} !== 'string'`);
    case SchemaType.Symbol:
      return fromNegation(`typeof ${valueAddress} !== 'symbol'`);
    case SchemaType.Test: {
      const testerVar = alloc("t", schema.tester);
      return fromNegation(`!${testerVar}.test(${valueAddress})`);
    }
    case SchemaType.Variant: {
      const isOneOf = getValidatorFromSchema(schema, keyAddress);
      const isOneOfVar = alloc("cv", isOneOf);
      return fromNegation(`!${isOneOfVar}(${valueAddress})`);
    }
    case SchemaType.Custom: {
      const customVar = alloc("custom", schema.customValidator);
      return fromNegation(`!${customVar}(${valueAddress})`);
    }
  }
}
