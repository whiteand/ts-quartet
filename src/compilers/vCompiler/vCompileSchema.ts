import { TSchema, CompilationResult } from "../../types";
import { getValidatorFromSchema } from "./getValidatorFromSchema";

export function vCompileSchema<T = any>(schema: TSchema): CompilationResult<T, any> {
  const explanations: any[] = [];
  const validator = getValidatorFromSchema<T>(schema);
  return Object.assign(validator, { explanations });
}
