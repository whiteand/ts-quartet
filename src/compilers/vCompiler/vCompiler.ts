import { RawSchema } from "../../IRawSchema";
import { rawSchemaToSchema } from "../../rawSchemaToSchema";
import { CompilationResult } from "../../types";
import { vCompileSchema } from "./vCompileSchema";

export function vCompiler<T = any>(
  rawSchema: RawSchema
): CompilationResult<T, any> {
  const schema = rawSchemaToSchema(rawSchema);
  return vCompileSchema<T>(schema);
}
