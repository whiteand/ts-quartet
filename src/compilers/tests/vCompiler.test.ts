import { anySchema, array, boolean, finite, neverSchema } from "../../schemas";
import { vCompiler } from "../vCompiler";
import { snapshot } from "./snapshot";

describe("v()", () => {
  test(`v(undefined)`, () => {
    snapshot(
      vCompiler,
      undefined,
      [undefined],
      [null, [undefined], "undefined", ["undefined"], 1, 0, NaN, false]
    );
  });
  test(`v(null)`, () => {
    snapshot(
      vCompiler,
      null,
      [null],
      ["null", [null], undefined, 0, 1, NaN, false]
    );
  });
  test(`v(true)`, () => {
    snapshot(
      vCompiler,
      true,
      [true],
      ["true", [true], 1, Infinity, NaN, false, null, undefined]
    );
  });
  test(`v(false)`, () => {
    snapshot(
      vCompiler,
      false,
      [false],
      ["false", [false], 1, Infinity, NaN, true, null, undefined]
    );
  });
  test(`v(0/0)`, () => {
    snapshot(
      vCompiler,
      0 / 0,
      [0 / 0],
      [1, "NaN", Infinity, -0, [0 / 0], "0/0"]
    );
  });
  test(`v(42)`, () => {
    snapshot(
      vCompiler,
      42,
      [42],
      [
        "42",
        [42],
        true,
        undefined,
        false,
        {},
        {
          valueOf() {
            return 42;
          }
        }
      ]
    );
  });
  test(`v('quartet')`, () => {
    snapshot(
      vCompiler,
      "quartet",
      ["quartet"],
      [
        ["quartet"],
        {
          toString() {
            return "quartet";
          }
        },
        1,
        null,
        NaN,
        undefined,
        Infinity
      ]
    );
  });
  test(`v('quartetquartetquartetquartetquartetquartetquartetquartetquartetquartetquartet')`, () => {
    snapshot(
      vCompiler,
      "quartetquartetquartetquartetquartetquartetquartetquartetquartetquartetquartet",
      [
        "quartetquartetquartetquartetquartetquartetquartetquartetquartetquartetquartet"
      ],
      [
        [
          "quartetquartetquartetquartetquartetquartetquartetquartetquartetquartetquartet"
        ],
        {
          toString() {
            return "quartetquartetquartetquartetquartetquartetquartetquartetquartetquartetquartet";
          }
        },
        1,
        null,
        NaN,
        undefined,
        Infinity
      ]
    );
  });
  test(`v(Symbol.for('quartet'))`, () => {
    snapshot(
      vCompiler,
      Symbol.for("quartet"),
      [Symbol.for("quartet")],
      [
        [Symbol.for("quartet")],
        "quartet",
        null,
        undefined,
        NaN,
        Infinity,
        false,
        true
      ]
    );
  });

  test("v(v.array)", () => {
    snapshot(
      vCompiler,
      array(),
      [[], [1], [[]]],
      [{ length: 10 }, null, undefined, false, "Array", Array]
    );
  });
  test("v(v.boolean)", () => {
    snapshot(
      vCompiler,
      boolean(),
      [true, false],
      [
        "true",
        "false",
        [true],
        [false],
        {
          toString() {
            return "true";
          }
        },
        null,
        undefined,
        0,
        NaN
      ]
    );
  });
  test("v(v.finite)", () => {
    snapshot(
      vCompiler,
      finite(),
      [1, 0, 1.5, -0],
      [NaN, Infinity, -Infinity, "0", "1", [1], [1.4], null, undefined, NaN]
    );
  });
  test("v(v.any)", () => {
    snapshot(
      vCompiler,
      anySchema(),
      [
        1,
        0,
        1.5,
        -0,
        NaN,
        Infinity,
        -Infinity,
        "0",
        "1",
        [1],
        [1.4],
        null,
        undefined,
        NaN
      ],
      []
    );
  });
  test("v(v.never)", () => {
    snapshot(
      vCompiler,
      neverSchema(),
      [],
      [
        1,
        0,
        1.5,
        -0,
        NaN,
        Infinity,
        -Infinity,
        "0",
        "1",
        [1],
        [1.4],
        null,
        undefined,
        NaN
      ]
    );
  });
  test("v(v.function)", () => {
    // TODO: Add v.function test
  });
  test("v(v.negative)", () => {
    // TODO: Add v.negative test
  });
  test("v(v.number)", () => {
    // TODO: Add v.number test
  });
  test("v(v.positive)", () => {
    // TODO: Add v.positive test
  });
  test("v(v.safeInteger)", () => {
    // TODO: Add v.safeInteger test
  });
  test("v(v.string)", () => {
    // TODO: Add v.string test
  });
  test("v(v.symbol)", () => {
    // TODO: Add v.symbol test
  });
});
