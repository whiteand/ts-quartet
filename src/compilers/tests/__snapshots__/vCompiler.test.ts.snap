// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`v() v('quartet') 1`] = `
Object {
  "code": "function (value) {
        return ctx.pure(value, ctx);
    }",
  "explanations": Array [],
  "memory": Object {
    "pathToValueId": Array [],
  },
  "pure": Object {
    "code": "function anonymous(value,ctx
) {
return value === \\"quartet\\"
}",
  },
}
`;

exports[`v() v('quartetquartetquartetquartetquartetquartetquartetquartetquartetquartetquartet') 1`] = `
Object {
  "code": "function (value) {
        return ctx.pure(value, ctx);
    }",
  "explanations": Array [],
  "memory": Object {
    "pathToValueId": Array [],
    "validString": "quartetquartetquartetquartetquartetquartetquartetquartetquartetquartetquartet",
  },
  "pure": Object {
    "code": "function anonymous(value,ctx
) {
return value === ctx.memory.validString
}",
  },
}
`;

exports[`v() v(0/0) 1`] = `
Object {
  "code": "function (value) {
        return ctx.pure(value, ctx);
    }",
  "explanations": Array [],
  "memory": Object {
    "pathToValueId": Array [],
  },
  "pure": Object {
    "code": "function anonymous(value,ctx
) {
return Number.isNaN(value)
}",
  },
}
`;

exports[`v() v(42) 1`] = `
Object {
  "code": "function (value) {
        return ctx.pure(value, ctx);
    }",
  "explanations": Array [],
  "memory": Object {
    "pathToValueId": Array [],
    "validNumber": 42,
  },
  "pure": Object {
    "code": "function anonymous(value,ctx
) {
return value === ctx.memory.validNumber
}",
  },
}
`;

exports[`v() v(Symbol.for('quartet')) 1`] = `
Object {
  "code": "function (value) {
        return ctx.pure(value, ctx);
    }",
  "explanations": Array [],
  "memory": Object {
    "pathToValueId": Array [],
    "validSymbol": Symbol(quartet),
  },
  "pure": Object {
    "code": "function anonymous(value,ctx
) {
return value === ctx.memory.validSymbol
}",
  },
}
`;

exports[`v() v(false) 1`] = `
Object {
  "code": "function (value) {
        return ctx.pure(value, ctx);
    }",
  "explanations": Array [],
  "memory": Object {
    "pathToValueId": Array [],
  },
  "pure": Object {
    "code": "function anonymous(value,ctx
) {
return value === false
}",
  },
}
`;

exports[`v() v(null) 1`] = `
Object {
  "code": "function (value) {
        return ctx.pure(value, ctx);
    }",
  "explanations": Array [],
  "memory": Object {
    "pathToValueId": Array [],
  },
  "pure": Object {
    "code": "function anonymous(value,ctx
) {
return value === null
}",
  },
}
`;

exports[`v() v(true) 1`] = `
Object {
  "code": "function (value) {
        return ctx.pure(value, ctx);
    }",
  "explanations": Array [],
  "memory": Object {
    "pathToValueId": Array [],
  },
  "pure": Object {
    "code": "function anonymous(value,ctx
) {
return value === true
}",
  },
}
`;

exports[`v() v(undefined) 1`] = `
Object {
  "code": "function (value) {
        return ctx.pure(value, ctx);
    }",
  "explanations": Array [],
  "memory": Object {
    "pathToValueId": Array [],
  },
  "pure": Object {
    "code": "function anonymous(value,ctx
) {
return value === undefined
}",
  },
}
`;

exports[`v() v(v.any) 1`] = `
Object {
  "code": "function (value) {
        return ctx.pure(value, ctx);
    }",
  "explanations": Array [],
  "memory": Object {
    "pathToValueId": Array [],
  },
  "pure": Object {
    "code": "function anonymous(value,ctx
) {
return true
}",
  },
}
`;

exports[`v() v(v.array) 1`] = `
Object {
  "code": "function (value) {
        return ctx.pure(value, ctx);
    }",
  "explanations": Array [],
  "memory": Object {
    "pathToValueId": Array [],
  },
  "pure": Object {
    "code": "function anonymous(value,ctx
) {
return Array.isArray(value)
}",
  },
}
`;

exports[`v() v(v.boolean) 1`] = `
Object {
  "code": "function (value) {
        return ctx.pure(value, ctx);
    }",
  "explanations": Array [],
  "memory": Object {
    "pathToValueId": Array [],
  },
  "pure": Object {
    "code": "function anonymous(value,ctx
) {
return typeof value === 'boolean'
}",
  },
}
`;

exports[`v() v(v.finite) 1`] = `
Object {
  "code": "function (value) {
        return ctx.pure(value, ctx);
    }",
  "explanations": Array [],
  "memory": Object {
    "pathToValueId": Array [],
  },
  "pure": Object {
    "code": "function anonymous(value,ctx
) {
return Number.isFinite(value)
}",
  },
}
`;

exports[`v() v(v.never) 1`] = `
Object {
  "code": "function (value) {
        return ctx.pure(value, ctx);
    }",
  "explanations": Array [],
  "memory": Object {
    "pathToValueId": Array [],
  },
  "pure": Object {
    "code": "function anonymous(value,ctx
) {
return false
}",
  },
}
`;
